File: EduSphereApp/edusphere-web/backend/controllers/adminController.js
const db = require('../config/db');

// Get system dashboard summary (admin overview)
exports.getDashboardStats = async (req, res) => {
  if (req.user.role !== 'admin') {
    return res.status(403).json({ message: 'Forbidden: Admin privilege required' });
  }

  try {
    const [userCount] = await db.query('SELECT COUNT(*) as count FROM users WHERE role != ?', ['admin']);
    const [completionsCount] = await db.query('SELECT COUNT(*) as count FROM topic_completions');
    const [quizzesCount] = await db.query('SELECT COUNT(*) as count FROM quiz_results');
    const [avgScore] = await db.query('SELECT AVG(percentage) as average FROM quiz_results');

    res.status(200).json({
      totalUsers: userCount[0].count,
      totalCompletions: completionsCount[0].count,
      totalQuizzes: quizzesCount[0].count,
      avgScore: avgScore[0].average ? Math.round(avgScore[0].average) : 0
    });
  } catch (error) {
    console.error('Admin dashboard stats error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

// List all registered students and their aggregate study progress
exports.getAllUsers = async (req, res) => {
  if (req.user.role !== 'admin') {
    return res.status(403).json({ message: 'Forbidden: Admin privilege required' });
  }

  try {
    const [users] = await db.query(
      `SELECT u.id, u.name, u.email, u.created_at, up.streak, up.total_completed, up.total_quizzes,
              CASE WHEN up.total_quizzes > 0 THEN ROUND(up.total_quiz_score / up.total_quizzes) ELSE 0 END as avg_score
       FROM users u
       LEFT JOIN user_progress up ON u.id = up.user_id
       WHERE u.role != 'admin'
       ORDER BY u.created_at DESC`
    );
    res.status(200).json(users);
  } catch (error) {
    console.error('Admin get all users error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

// Delete a user (moderation)
exports.deleteUser = async (req, res) => {
  if (req.user.role !== 'admin') {
    return res.status(403).json({ message: 'Forbidden: Admin privilege required' });
  }

  const { userId } = req.params;

  try {
    const [result] = await db.query('DELETE FROM users WHERE id = ? AND role != ?', [userId, 'admin']);
    if (result.affectedRows === 0) {
      return res.status(404).json({ message: 'User not found or cannot delete an admin' });
    }
    res.status(200).json({ message: 'User and all related records deleted successfully' });
  } catch (error) {
    console.error('Admin delete user error:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

File: EduSphereApp/edusphere-web/backend/src/main/java/com/edusphere/controller/AdminController.java
package com.edusphere.controller;

import com.edusphere.config.JwtAuthFilter.UserPrincipal;
import com.edusphere.dto.AdminStatsResponse;
import com.edusphere.model.User;
import com.edusphere.model.UserProgress;
import com.edusphere.repository.*;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/admin")
@PreAuthorize("hasRole('ADMIN')")
public class AdminController {

    private final UserRepository userRepository;
    private final TopicCompletionRepository topicCompletionRepository;
    private final QuizResultRepository quizResultRepository;
    private final UserProgressRepository userProgressRepository;

    public AdminController(
            UserRepository userRepository,
            TopicCompletionRepository topicCompletionRepository,
            QuizResultRepository quizResultRepository,
            UserProgressRepository userProgressRepository
    ) {
        this.userRepository = userRepository;
        this.topicCompletionRepository = topicCompletionRepository;
        this.quizResultRepository = quizResultRepository;
        this.userProgressRepository = userProgressRepository;
    }

    @GetMapping("/stats")
    public ResponseEntity<AdminStatsResponse> getDashboardStats() {
        long totalUsers = userRepository.count();
        long totalCompletions = topicCompletionRepository.count();
        long totalQuizzes = quizResultRepository.count();

        List<UserProgress> allProgress = userProgressRepository.findAll();
        int totalScore = 0;
        int totalQuizCount = 0;
        for (UserProgress up : allProgress) {
            if (up.getTotalQuizzes() != null && up.getTotalQuizzes() > 0) {
                totalQuizCount += up.getTotalQuizzes();
                totalScore += up.getTotalQuizScore();
            }
        }
        int avgScore = totalQuizCount > 0 ? Math.round((float) totalScore / totalQuizCount) : 0;

        return ResponseEntity.ok(new AdminStatsResponse(totalUsers, totalCompletions, totalQuizzes, avgScore));
    }

    @GetMapping("/users")
    public ResponseEntity<List<Map<String, Object>>> getAllUsers() {
        List<User> users = userRepository.findAll();
        List<Map<String, Object>> result = new ArrayList<>();

        for (User u : users) {
            Map<String, Object> map = new HashMap<>();
            map.put("id", u.getId());
            map.put("name", u.getName());
            map.put("email", u.getEmail());
            map.put("role", u.getRole());
            map.put("created_at", u.getCreatedAt());

            Optional<UserProgress> progressOpt = userProgressRepository.findByUserId(u.getId());
            if (progressOpt.isPresent()) {
                UserProgress up = progressOpt.get();
                map.put("streak", up.getStreak());
                map.put("total_completed", up.getTotalCompleted());
                map.put("total_quizzes", up.getTotalQuizzes());
            }
            result.add(map);
        }
        return ResponseEntity.ok(result);
    }

    @DeleteMapping("/user/{userId}")
    public ResponseEntity<?> deleteUser(@PathVariable Long userId) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            Map<String, String> error = new HashMap<>();
            error.put("message", "User not found");
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
        }

        // Prevent self-deletion
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        if (principal instanceof UserPrincipal) {
            Long currentId = ((UserPrincipal) principal).getId();
            if (currentId.equals(userId)) {
                Map<String, String> error = new HashMap<>();
                error.put("message", "Cannot delete your own account");
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
            }
        }

        userRepository.deleteById(userId);
        Map<String, String> body = new HashMap<>();
        body.put("message", "User deleted successfully");
        return ResponseEntity.ok(body);
    }
}
